### PR TITLE
adding code for masking and cleaned/masked fine-tuning data

### LINKED ISSUE #17: Fine-Tuning the CodeT5 Model for R Language in Bioinformatics
The CodeT5 model needs to be fine-tuned using bioinformatics-specific R datasets to enhance its code generation, debugging, and optimization capabilities. The fine-tuned model should understand Bioconductor functions commonly used in bioinformatics research.

Acceptance Criteria:
- Preprocess the collected dataset for fine-tuning.
- Fine-tune the CodeT5 model using the curated R language datasets.

### PR DESCRIPTION
**Pull request by PranaviKolouju**
Fixes #17 

**What was changed?**

I created a script to combine the three R datasets that we perviously collected from bioconductor, R packages, and online tutorials to prepare them for finetuning. In addition to combining these datasets, I also masked each of the code snippets to create input data and output labels and saved these out to a file. 

**Why was it changed?**

In order to finetune the CodeT5 model for code generation tasks on R code, our dataset needs to be structured in such a way where the input data is the full code snippet with some of the code removed and the training label is the removed bit of code. So, I performed a masking operation on each code snippet that structures our dataset in the previously described format. 

**How was it changed?**

I added a jupyter notebook titled `mask_r_code.ipynb` that combines all three datasets that were previously collected (bioconductor, R packages, and online tutorials). It also performs a masking operation using the hole-punch methodology where either one line, multiple lines, or a substring of characters is removed from the code snippet. Then, the remaining code becomes the training example and the removed code becomes the training label. The input and label are saved in a file titled `fixed_masked_training_data.json` labeled as either "input" or "output".

### CODE REVIEW

File: Fine Tuning/masking/mask_r_code.ipynb [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Generated 3599 hole-punched training samples.\n"
```
Comment (Omp06) [inline Fine Tuning/masking/mask_r_code.ipynb:114]: Does the hole-punching strategy handle cases where there are very few lines of code properly?

File: Fine Tuning/masking/mask_r_code.ipynb [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    "                    continue \n",
+    "\n",
+    "            elif strategy == \"multi_line_removal\" and len(lines) > 3:\n",
+    "                start_idx = random.randint(0, len(lines) - 3)\n",
+    "                lines_to_remove = lines[start_idx : start_idx + random.randint(2, 3)]\n",
+    "                masked_code = code.replace(\"\\n\".join(lines_to_remove), \"____\", 1)\n",
+    "                masked_data.append({\"input\": masked_code, \"output\": \"\\n\".join(lines_to_remove)})\n",
+    "                continue\n",
+    "\n",
+    "        words = code.split()\n",
```
Comment (Omp06) [inline Fine Tuning/masking/mask_r_code.ipynb:151]: Could replacing a random snippet create cases where variable names or function calls are broken?

File: Fine Tuning/masking/mask_r_code.ipynb [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+   },
+   "file_extension": ".py",
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.12.1"
+  }
+ },
+ "nbformat": 4,
```
Comment (Omp06) [inline Fine Tuning/masking/mask_r_code.ipynb:188]: I feel that this script does very well with the masking job. My only concern would be if we are doing it in the most optimal way possible for the CodeT5 model. Is a possibility that we might need to clean the data itself before we mask it?

File: Fine Tuning/masking/mask_r_code.ipynb [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import json\n",
+    "\n",
+    "with open(\"bioinformatics_workshop_gitbook.json\", \"r\", encoding=\"utf-8\") as file:\n",
+    "    data = json.load(file)\n",
+    "\n",
```
Comment (AjithAkuthota23) [inline Fine Tuning/masking/mask_r_code.ipynb:13]: Does this load completely for the larger json files? If not, consider streaming the json files

File: Fine Tuning/masking/mask_r_code.ipynb (hunk 2 above)
Comment (AjithAkuthota23) [reply Fine Tuning/masking/mask_r_code.ipynb:151]: Yeah, Does this happen for the smaller functions?

File: Fine Tuning/masking/mask_r_code.ipynb [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    "                valid_lines = [line for line in lines if not line.strip().startswith(\"#\")]\n",
+    "                if valid_lines:\n",
+    "                    line_to_remove = random.choice(valid_lines)\n",
+    "                    masked_code = code.replace(line_to_remove, \"____\", 1)\n",
+    "                    masked_data.append({\"input\": masked_code, \"output\": line_to_remove})\n",
+    "                    continue \n",
+    "\n",
+    "            elif strategy == \"multi_line_removal\" and len(lines) > 3:\n",
+    "                start_idx = random.randint(0, len(lines) - 3)\n",
+    "                lines_to_remove = lines[start_idx : start_idx + random.randint(2, 3)]\n",
```
Comment (AjithAkuthota23) [inline Fine Tuning/masking/mask_r_code.ipynb:146]: Can we exclude some of the key words like def, return?

File: Fine Tuning/masking/mask_r_code.ipynb (hunk 4 above)
Comment (PranaviKolouju) [reply Fine Tuning/masking/mask_r_code.ipynb:13]: It does!

File: Fine Tuning/masking/mask_r_code.ipynb (hunk 5 above)
Comment (PranaviKolouju) [reply Fine Tuning/masking/mask_r_code.ipynb:146]: For larger functions, it extract a whole line so words like "def" would be masked along with them.

**Review by AjithAkuthota23 — APPROVED** [review]
